contracts/weapons.sol

```diff
@@ -379,52 +382,92 @@ contract Weapons is Initializable, ERC721Upgradeable, AccessControlUpgradeable,
         return result;
     }
 
+    function getDustSupply(uint8 stars) public view returns (uint16) {
+        return uint16(burnDust[msg.sender] >> (stars * 16));
+    }
+
+    function burn(uint256 burnID) public restricted {
+        Weapon storage burning = tokens[burnID];
+        uint8 stars = getStarsFromProperties(burning.properties);
+        require(getDustSupply(stars) < 65535, "Dust supply capped");
+        _burn(burnID);
+        burnDust[msg.sender] += (uint256(1) << (stars * 16));
+
+        emit Burned(
+            ownerOf(burnID),
+            burnID
+        );
+    }
+
     function reforge(uint256 reforgeID, uint256 burnID) public restricted {
-        WeaponBurnPoints storage wbp = burnPoints[reforgeID];
         Weapon storage burning = tokens[burnID];
+        uint8 stars = getStarsFromProperties(burning.properties);
+        applyBurnPoints(reforgeID, stars);
+        _burn(burnID);
+
+        WeaponBurnPoints storage wbp = burnPoints[reforgeID];
+        emit Reforged(
+            ownerOf(reforgeID),
+            reforgeID,
+            burnID,
+            wbp.lowStarBurnPoints,
+            wbp.fourStarBurnPoints,
+            wbp.fiveStarBurnPoints
+        );
+    }
 
-        if(getStarsFromProperties(burning.properties) == 0) { // 1 star
+    function reforgeWithDust(uint256 reforgeID, uint8 stars) public restricted {
+        require(getDustSupply(stars) > 0, "Dust supply needed");
+        applyBurnPoints(reforgeID, stars);
+        burnDust[msg.sender] -= (uint256(1) << (stars * 16));
```

Comment:
Please move this and the `require()` statement just above that checks dust supply, to a dedicated `internal` function called something like `_decrementDustSupply(address playerAddress, uint8 stars)`. We want to make sure to encapsulate working with burn dust, especially when bit operations like this are involved (yes, even if the code seems simple).